core-schema/src/schema.rs

```diff
@@ -0,0 +1,585 @@
+            .collect();
+
+        let using_features = match bootstrap(&schema) {
+            Some(req) => req,
+            None => {
+                errors.push(SchemaError::NoCore);
+                return Ok(Schema {
+                    document,
+                    features: vec![],
+                    errors,
+                });
+            }
+        };
+
+        let mut using = vec![];
+
+        for (dir, req) in features.drain(0..) {
+            if dir.name != using_features.name {
+                continue;
+            }
+            match req {
+                Ok(req) => using.push(req),
+                Err(err) => errors.push(SchemaError::BadUsingRequest(dir.position, err)),
+            }
+        }
+
+        let mut machined = Schema {
+            document,
+            features: using,
+            errors,
+        };
+        machined.validate_no_overlapping_prefixes();
+        Ok(machined)
+    }
+
+    /// Validate that no two `Feature`s are using the same `prefix`, removing
+    /// *all* overlapping `Feature`s from `self.features` and reporting
+    /// `SchemaError::OverlappingPrefix` in `self.errors`.
+    fn validate_no_overlapping_prefixes(&mut self) {
+        let mut by_prefix = HashMap::<_, u32>::new();
+        for req in &self.features {
+            let count = by_prefix.entry(req.name.clone()).or_default();
+            *count += 1;
+        }
+        for (prefix, count) in by_prefix.drain() {
+            if count > 1 {
+                self.errors.push(SchemaError::OverlappingPrefix(
+                    prefix.clone(),
+                    drain_filter_collect(&mut self.features, |req| req.name == *prefix),
+                ));
+            }
+        }
+    }
+
+    /// Given a set of implementations, take all features in the document and
+    /// match them to the implementations which satisfy them ("satisfying
+    /// implementations").
+    ///
+    /// This returns an `Iterator` over `(&Feature, Find<T>)`.
+    /// `Find` is an `Iterator` over `(&Version, &T)`, which will iterate satisfying
+    /// implementations ordered by lowest to highest version. The
+    /// [`bound`](Bounds.html#bound) method can be used to get the lowest and highest
+    /// satisfying versions.
+    ///
+    /// # Example:
+    ///
+    /// ```
+    /// use core_schema::*;
+    ///
+    /// let impls = Implementations::new()
+    ///     .provide("https://spec.example.com/A", Version(1, 2), "impl for A 1.2")
+    ///     .provide("https://spec.example.com/A", Version(1, 3), "impl for A 1.3")
+    ///     .provide("https://spec.example.com/A", Version(1, 7), "impl for A 1.7");
+    /// let schema = Schema::parse(r#"schema @core(feature: "https://lib.apollo.dev/core/v0.1") @core(feature: "https://spec.example.com/A/v1.0") { query: Query }"#)?;
+    ///
+    /// let max = schema
+    ///   .activations(&impls)
+    ///   .filter_map(|(_request, mut impls)| impls.bounds())
```

Comment:
this impls is a different type than the impls on line 237 which I find confusing